### DISCUSSION TITLE
Read an existing file on the filesystem without copying

### DISCUSSION THREAD

**Discussion by gmaclennan** [rhashimoto/wa-sqlite#226] [Q&A] [answered]
I am trying to understand how I might use this to read an existing sqlite database in the user filesystem without copying the database into OPFS. I am trying to build an offline web app for reading sqlite files (specifically, [mbtiles files](https://github.com/mapbox/mbtiles-spec)). The files can be large (> 1GB) so I want to avoid reading them into memory.

In an ideal world, we could use [`window.showOpenFilePicker()`](https://developer.mozilla.org/en-US/docs/Web/API/Window/showOpenFilePicker) to return a `FileSystemFileHandle` and then call `createSyncAccessHandle()` to get a [`FileSystemSyncAccessHandle`](https://developer.mozilla.org/en-US/docs/Web/API/FileSystemSyncAccessHandle), which I could then pass to a VFS, but I'm not sure which one?

For supporting browsers that don't support `window.showOpenFilePicker()` (unfortunately, anything other than Chrome), then it seems like I could maybe use `OPFSAnyContextVFS` to read a [`File`](https://developer.mozilla.org/en-US/docs/Web/API/File) object from opening a file selected from a user input, but I'm unclear about how I could pass the `File` object to initialize the database?

The alternative is to streaming copy a `File` object into OPFS, and then use an OPFS VFS to read the copied file, although I'm not entirely sure which one to use.

I understand that SQLite needs to create other files for writing, but for just reading a database can it just use a single file handle to read the DB file?

**Comment by rhashimoto** [comment, marked as answer]
> The alternative is to streaming copy a File object into OPFS, and then use an OPFS VFS to read the copied file, although I'm not entirely sure which one to use.

This will work. You could use any OPFS VFS with filesystem transparency (see [feature table](https://github.com/rhashimoto/wa-sqlite/tree/master/src/examples#vfs-comparison)). The drawback is the time to copy a big file.

If you want to access the user's file directly (without copying to OPFS), you'll need to write a custom VFS. I'm unsure whether you can get a FileSystemSyncAccessHandle on a picked file (i.e. outside OPFS) if you're not on a Chromium browser; my guess is that Firefox and Safari don't support it. Also note that FileSystemWritableFileStream isn't supported on Safari and Android Chrome.

If you only want to read the user's database file then your custom VFS might work with a simple modification to OPFSAnyContextVFS. You could try replacing this code:

https://github.com/rhashimoto/wa-sqlite/blob/605100a88caa62e5ab7f7c5f4b453b8c6c016bab/src/examples/OPFSAnyContextVFS.js#L72-L74

with something like:

```javascript
  file.fileHandle = getUsersFileSystemFileHandle(zName);
```

where you would implement `getUsersFileSystemFileHandle()` to return the user's file handle or throw if any other file was requested.

> I understand that SQLite needs to create other files for writing, but for just reading a database can it just use a single file handle to read the DB file?

Mostly yes, but it will generally still attempt to open other files to check that the database wasn't left in an inconsistent state. In the typical case these other files will be not present or empty.

**Reply by subframe7536** [reply]
It has performance issue on Windows... Haven't test on other platforms

https://github.com/subframe7536/sqlite-wasm/pull/5

**Reply by rhashimoto** [reply]
OPFSAnyContextVFS writes are expected to be very slow. Discussion [here](https://github.com/rhashimoto/wa-sqlite/discussions/184):

> Reading with File is slower than an access handle, but only moderately slower - you might make up the performance by avoiding round-trip messaging with a Worker. Writing with FileSystemWritableFileStream, on the other hand, is horribly slow - you actually write to a temporary copy that is scanned for malware and then renamed to replace the original file. So even the smallest edit requires reading the entire file twice (once to copy, once for malware scan).

**Comment by gmaclennan** [comment]
Thank you @rhashimoto this is really helpful. Still investigating the best way to do this - the answer might just be "wait a year for better browser support".
